src/riscv-integration.adoc

```diff
@@ -1102,3 +1124,17 @@ written to these CSRs unless the specification explicitly states that the CSR
 behaves otherwise (see <<mepcc>> and <<sepcc>>). Also note that <<pcc>> can
 be written with a <<JALR>> instruction in Capability Mode which automatically unseal the capability _before_
 the invalid address conversion above.
+
+[#CHERI_COMP,reftext="CHERI Compartmentalization]
+=== CHERI Compartmentalization
+
+This section describes how this specification enables support for compartmentalization for CHERI systems.
+Compartmentalization seeks to separate the privileges between different protection units, e.g., two or more libraries.
+Code can be separated by sentries, which allow for giving out code capabilities to untrusted code where the untrusted code can only call code capability, but not modify it.
+Sentries can be called from different threads and thus there needs to be a way of identifying the current thread.
+While identifying the current thread can be done by privileged code, e.g., the kernel, the implied performance overhead of this is not bearable for CHERI systems with many compartments.
+
+The RISC-V ABI includes a _thread pointer (tp)_ register, which is not usuable for the purpose of reliably identifying the current thread because the tp register is a general purpose register and can be changed arbitrarily by untrusted code.
+Therefore, this specification offers two additional CSRs that facilitate a trusted source for the thread ID.
+The supervisor thread identifier (STID) register is readable and writeable with <<asr_perm>> from the supervisor mode allowing to change the thread ID on a context switch.
+The user thread identifier (UTID) exposes the current value of STID as a read-only copy.
```

Comment:
@andresag01:

A compartment transition stays within a software thread. Usually, there is a trampoline in between a compartment transition in order to facilitate building up a data capability for the called compartment. A call from "CompA" to "CompB" looks as follows:

CompA -> Trampoline -> CompB

After "CompB" has finished the task it will return to "CompA", which also goes through trampoline:

CompB -> Trampoline -> CompA

All transitions are using sentries. The trampoline needs to have access to a trusted stack in order to find its own information as well as finding the data capability of the compartment it is transitioning to. The TID is needed in order to identify which trusted stack is the correct one. The same trampoline code can be invoked from different threads. Imagine two threads doing exactly the same thing, e.g., making a call from "CompA" to "CompB". The trampoline needs to use a different trusted stack for each thread it is invoked from.

The user code does not need to change the thread ID - that is why `utid` is read-only. The kernel changes the `stid` register whenever it changes the software thread running on a hardware thread.

Reply:
Ok, thanks I see what you mean. My description is a little confused regarding the meaning of "compartment" and "thread".

In previous discussion, you mentioned that the thread scheduling could be done in user or supervisor mode. If so, then this stid/utid arrangement is only for supervisor mode. Correct? Did I misunderstand something?

Reply:
There can be multiple levels of thread scheduling. When there is user-level threading, there will also be supervisor-level threading. `stid` and `utid` allow for the thread ID to be communicated from supervisor mode to user mode. We currently believe that there is a way to provide a thread id within a privilege domain via read-only capabilities.